accounts-db/src/accounts.rs

```diff
@@ -753,18 +756,24 @@ impl Accounts {
                 .filter(|(i, _)| is_storable_account(message, *i))
             {
                 if message.is_writable(i) {
-                    let is_fee_payer = i == 0;
-                    let is_nonce_account = prepare_if_nonce_account(
-                        address,
-                        account,
-                        execution_status,
-                        is_fee_payer,
-                        maybe_nonce,
-                        durable_nonce,
-                        lamports_per_signature,
-                    );
-
-                    if execution_status.is_ok() || is_nonce_account || is_fee_payer {
+                    let should_store_account = match tx_status {
+                        TransactionStatus::Successful => true,
+                        TransactionStatus::FailedWithNonce { nonce } => {
+                            let is_fee_payer = i == 0;
+                            let is_nonce_account = prepare_if_nonce_account(
```

Comment:
Since `prepare_if_nonce_account` always exited early if `maybe_nonce` was `None`, we can refactor to only call it for failed nonce transactions where the `nonce` is available.